### PR TITLE
[dmd-cxx] fix Issue 19912: No implicit import of object module when an object declaration exists.

### PR DESCRIPTION
**Pull request by ibuclaw**

### CODE REVIEW

**Review by WalterBright — APPROVED** [review]

**Comment by JinShil** [comment]
Are you all sure about this?  It seems that instead, any `object` declaration would generate an error just like it does for `TypeInfo`  See https://run.dlang.io/is/1mwpVi  That being said, I don't like the fact that these runtime objects *must* be declared in `object.d`, but it seems to be a more systemic issue than just `object`.

I'd prefer that `object.d` were not automatically imported at all, but [I tried that], and was overruled.

Another option I'm considering is renaming `object.d` to `package.d` and have the compiler automatically import the runtime "package" instead of a specific module.

Can't we just have `dmd.conf` declare the automatic import of `object.d` or something so it's not hard-coded into the compiler like it currently is?

Anyway, can we discuss how to move forward on this?  I think there's more to consider here than just what this PR addresses.

**Comment by thewilsonator** [comment]
> I'd prefer that object.d were not automatically imported at all, but I tried that, and was overruled.

I think a better idea would be to have an auto imported package that would by default import object, that way we could split up the Frankenstein that is `object.d`. Then you could easily remove stuff that you don't want, since if you are going to the effort of not wanting classes deleting a file from the runtime that you are using is a lot easier than ripping out bits of `object.d`.

**Comment by JinShil** [comment]
> It seems that, instead, any object declaration would generate an error just like it does for TypeInfo See https://run.dlang.io/is/1mwpVi 

Oops, sorry.  I see now that this PR is actually adding that behavior.  That is consistent with the status quo, so 👍 

I think this PR is OK as is, as it maintains the status quo, but there's still a systemic problem with the way D automatically imports the runtime and the fact that it requires certain symbols to be declared in a specific place.  But, that's something that can be addressed later, I suppose.

**Review by JinShil — APPROVED** [review]

**Comment by ibuclaw** [comment]
Yes, this just makes it so that you can't override the automatic import with another declaration of the same name.

File: test/fail_compilation/fail19912b.d [hunk 1]
```diff
@@ -0,0 +1,9 @@
+// PERMUTE_ARGS:
+/*
+TEST_OUTPUT:
+---
+fail_compilation/fail19912b.d(8): Error: class `fail19912b.object` conflicts with import `fail19912b.object` at fail_compilation/fail19912b.d
```
Comment (jacob-carlborg) [inline test/fail_compilation/fail19912b.d]: The error message mentions `import`, but there is `import` in the source code. Can we come up with a better message?

**Pull request merged**